Move to only CSS-based approach - RAC completion - Removal of SCSS (#5766)
plone/volto: packages/components/README.md

Alternatively, as RAC allows, you can also drop your own basic set of styles.
You can take the basic styles as reference.
You can even bring your own CSS framework and make basic get the styling in there using the utilities of your choice.
It's even possible to use TailwindCSS for styling the components in this package, using the RAC styling approach.

## Components list

### Basic

- Button
- Checkbox
- CheckboxField
- CheckboxGroup
- Combobox
- Container
- Dialog
- Form
- GridList
- Icon
- Link
- Listbox
- Menu
- Modal
- NumberField
- Popover
- RadioGroup
- Select
- Switch
- Tabs
- TagGroup
- TextField
- TextAreaField
- ToggleButton
- Toolbar
- BlockToolbar
- Tooltip

### Viewlets

- Breadcrumbs

### Views

- DefaultBlockView
- RenderBlocks

### Quanta

- TextField
- TextAreaField

## Quanta icons

This package provides an implementation of the Quanta Icon set in React components.
They can be used directly in your components as:

```tsx
import { ChevronupIcon } from '../Icons/ChevronupIcon';

const MyComponent = (props) => (
  <Button aria-label="Unfold/Collapse">

>     {props.isOpen ? <ChevronupIcon /> : <ChevrondownIcon />}

Reviewer: @sneridagh I know this is just demo example code but there is no import of ChevrondownIcon,
or where Button comes from. We can keep examples to minimum but I still think they should be
as close to working as possible which to me means that this example is missing the Button and
the ChevrondownIcon import

  </Button>
)
```

## Helper providers

### `flattenToAppURL`

This provider allows you to pass down to your components a `flattenToAppURL` helper.
This helper can be different across platforms or frameworks, since the way to provide configuration can also be different.

```tsx
import { FlattenToAppURLProvider } from '@plone/components';
import { flattenToAppURL } from './utils';

const rootApp = (
  <FlattenToAppURLProvider flattenToAppURL={flattenToAppURL}>
    {children}
  </FlattenToAppURLProvider>
)
```

Then from your components:

```tsx
import { useFlattenToAppURL } from '@plone/components';

const Link = (props: LinkProps, ref: ForwardedRef<HTMLAnchorElement>) => {
  const { flattenToAppURL } = useFlattenToAppURL();
  const flattenedURL = flattenToAppURL(props.href);

  return (

>     <RACLink ref={ref} {...props} href={flattenedURL}>

Reviewer: @sneridagh I would add the import for RACLink in this example

> import { Link, useFlattenToAppURL } from '@plone/components';
> 
> const Link = (props: LinkProps, ref: ForwardedRef<HTMLAnchorElement>) => {
>   const { flattenToAppURL } = useFlattenToAppURL();
>   const flattenedURL = flattenToAppURL(props.href);
> 
>   return (
>     <Link {...props} href={flattenedURL}>
>       {props.children}
>     </RACLink>
>   );
> };

Reviewer: ```suggestion
import { useFlattenToAppURL } from '@plone/components';
import { Link as RACLink } from 'react-aria-components';

const Link = (props: LinkProps, ref: ForwardedRef<HTMLAnchorElement>) => {
  const { flattenToAppURL } = useFlattenToAppURL();
  const flattenedURL = flattenToAppURL(props.href);

  return (
    <RACLink {...props} href={flattenedURL}>
      {props.children}
    </RACLink>
  );
};
```

Author: ok, but both could work!

- Agnostic (not tied to any known design system)
- Composable (they are ready to be the building blocks of other complex components)
- Theme-able (they can be themed, using both the provided basic theme as a baseline and custom CSS properties)
- Data-driven-less (they are dumb, presentational components)
- i18n-less (they do not provide i18n machinery or attached to any i18n framework or library)
- Built on a renowned headless CMS components library: React Aria Components (RAC) (https://react-spectrum.adobe.com/react-aria/components.html)

Since we are using RAC as the base, the styling of this package components are done via the RAC styling options (https://react-spectrum.adobe.com/react-aria/styling.html).
See the `react-aria-components` section below for more information about RAC.
This package provides a simple and basic set of CSS rules to build upon.
Alternatively, you can bring your own styles, removing the basic styling or complementing it, and build on top of it.
You can even use RAC to use other CSS utility libraries, like TailwindCSS.

## Styling

This package provide a basic set of CSS rules.

> We call them "BasiQ" (the "Q" is for "Quanta").

Reviewer: @sneridagh @stevepiercy  No one asked my opinion but here I am giving it anyhow :) 
I am not so sure if it would be a good thing to have a name that is like an acronym.
I am thinking of how we are going to spell it when we give training about it.
Personally I wouldn't try to make clever names in trying to combine something
and giving it meaning then somehow.

I also think we should investigate if the naming doesn't clash with other  entities that use
this naming, performing a google search gave me a few results for basiq such as 
https://www.basiq.io/
I don't know how far are trademarks and copyright nowadays but it doesn't hurt for us todo
some searching and ensuring we don't use something already established and copyrighted

Reviewer 2: I had a similar moment of "hmmm", how do you even pronounce that? "Bay-sik" or "Bay-suh-que"?

"Basiq" is trademarked.

Let's avoid the attorney horde. I'll push a commit to replace it with "Quanta base CSS rules" and grammatically correct phrases.

Reviewer 2: On second thought, I won't push a commit. It means renaming the package.

Author: @stevepiercy 
I've already talked about this with @ichim-david 
I came up with the name in 5mins. It's an internal name for now, we won't use it anywhere else. 
We will sell `@plone/components` what's inside stays in Vegas.
Let's maintain it as internal name until someone has a better name. And yes, we need a name for them.
Quanta is also something else out there, nobody complained.

Author: @stevepiercy 
OTOH, I after giving it a second thought, change a Q for a c is easy anyways, but let's all think in the name of the basic/fundamental/foundational CSS rules.

Reviewer 2: @sneridagh indeed, I think clever spelling in names end up being a footgun. I think it is better not to brand something that you do not intend to put much marketing effort toward. It's just confusing to developers.

Author: Ready.

```

Take a look at the implementation of the Quanta components, using the BasiQ ones as baseline in the `quanta` folders.

Alternatively, as RAC allows, you can also drop your own basic set of styles.
You can take the BasiQ styles as reference.
You can even bring your own CSS framework and make BasiQ get the styling in there using the utilities of your choice.
It's even possible to use TailwindCSS for styling the components in this package, using the RAC styling approach.

## Components list

### BasiQ

- Button
- Checkbox
- CheckboxField
- CheckboxGroup
- Combobox
- Container
- Dialog
- Form
- GridList
- Icon
- Link
- Listbox
- Menu
- Modal
- NumberField
- Popover
- RadioGroup
- Select
- Switch
- Tabs
- TagGroup
- TextField
- TextAreaField
- ToggleButton
- Toolbar
- BlockToolbar
- Tooltip

### Viewlets

- Breadcrumbs

### Views

- DefaultBlockView
- RenderBlocks

### Quanta

- TextField
- TextAreaField

## Quanta icons

This package provides an implementation of the Quanta Icon set in React components.
They can be used directly in your components as:

```tsx
import { ChevronupIcon } from '../Icons/ChevronupIcon';

const MyComponent = (props) => (

>   <Button>

Author: I added `aria-label` on the icons. Question is if we want them to provide one by default or not. Is that what you meant?

>   <Button>

Reviewer 3: I have to admit I did not notice the `aria-label` on the icons, sorry. I am not entirely sure it's enough though, for these reasons:

- `title` also provides a tooltip which is nicer for visual users that only see an icon instead (even if totally understandable like a chevron in this case)
- iirc the title or label should be on the button, because screen readers will read that as the button label and it is more straightforward to understand for the user. Also, the icon itself usually has `aria-hidden="true"` so I'm not sure if the label would be read, I haven't tested it.

To answer you question, I only meant to have a good example in the docs, I did not mean to change the components code.

>   <Button>

Author: OK! fortunately RAC will help with all these, since they already cover a lot, but these guidelines will be great to have sumarized somewhere (probably in the components docs too).

>   <Button>

Author: Btw, `title` is not in the RAC Button props... that is indicator that having the title is no longer the best practice or the browsers have now support for aria instead...

>   <Button>

Author: https://react-spectrum.adobe.com/react-spectrum/Button.html#accessibility

>   <Button>

Author: Also on that regard, I plan to make a test on A11y with AXE for all components.

>   <Button>

Reviewer 3: > Btw, `title` is not in the RAC Button props... that is indicator that having the title is no longer the best practice or the browsers have now support for aria instead...

Yes, this is correct. Also, `aria-label` always takes precedence when set, it's the best way to go. What I kind of dislike of this new spec is that `title` also added the tooltip and I really don't get why that is not an important feature for people making decisions on these topics :D anyway, I'm happy if we stick to `aria-label` here

You can override them in your classes while maintaining them for others.

### Quanta

This package also features the Quanta components.
The Quanta theme is an example of it.
These components use the basic styling as a baseline, not only in styling, but also in the component side, reusing the CSS and custom CSS properties in it.

Quanta is built upon the basic styles in an additive way.
The use of the Quanta CSS implies using it upon basic styling.
You could take Quanta as example to build your own layer of styles over basic styling for your theme.

To use a theme built upon the basic styling, you need to import both the basic and the theme CSS, in this order:

```js
import '@plone/components/basic.css';
import '@plone/components/quanta.css';
```

You have the option of doing it selectively per component, too:

```js
import '@plone/components/src/styles/basic/TextField.css';
import '@plone/components/src/styles/quanta/TextField.css';
```

Take a look at the implementation of the Quanta components, using the basic ones as baseline in the `quanta` folders.

Alternatively, as RAC allows, you can also drop your own basic set of styles.
You can take the basic styles as reference.

> You can even bring your own CSS framework and make basic get the styling in there using the utilities of your choice.

Reviewer 2: ```suggestion
You can even bring your own CSS framework, and add your styles to the basic styles using the utilities of your choice.
```

Author: Not really the intention... although valid, we can leave it as well. The initial intention is that you can remove them completely, and use your framework to style from scratch the default RAC classNames...

Author: See last commit.
